src/test/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporterTest.java

```diff
@@ -0,0 +1,513 @@
+                .timeoutMs(100000L)
+                .build();
+
+        target = new GreenbidsAnalyticsReporter(
+                greenbidsAnalyticsProperties,
+                jacksonMapper,
+                httpClient,
+                clock,
+                prebidVersionProvider);
+
+        jsonCaptor = ArgumentCaptor.forClass(String.class);
+    }
+
+    @Test
+    public void shouldReceiveValidResponseOnAuctionContextForBanner() throws IOException {
+        // given
+        final Banner banner = setUpBanner();
+
+        final ObjectNode prebidJsonNodes = mapper.valueToTree(
+                singletonMap("gpid", TextNode.valueOf("gpidvalue")));
+
+        final Imp imp = Imp.builder()
+                .id("adunitcodevalue")
+                .ext(prebidJsonNodes)
+                .banner(banner)
+                .build();
+
+        final AuctionContext auctionContext = setupAuctionContext(imp);
+        event = AuctionEvent.builder()
+                .auctionContext(auctionContext)
+                .bidResponse(auctionContext.getBidResponse())
+                .build();
+
+        final HttpClientResponse mockResponse = mock(HttpClientResponse.class);
+        when(mockResponse.getStatusCode()).thenReturn(202);
+        when(httpClient.post(anyString(), any(MultiMap.class), jsonCaptor.capture(), anyLong()))
+                .thenReturn(Future.succeededFuture(mockResponse));
+
+        // when
+        final Future<Void> result = target.processEvent(event);
+
+        final String capturedJson = jsonCaptor.getValue();
+        final ObjectNode capturedJsonNode = (ObjectNode) mapper.readTree(capturedJson);
+        capturedJsonNode.put("greenbidsId", "testGreenbidsId");
+        capturedJsonNode.put("billingId", "testBillingId");
+        final ObjectNode expectedJsonNode = (ObjectNode) mapper.readTree(parseJson(EXPECTED_JSON_PATH_BANNER));
+
+        // then
+        assertThat(result.succeeded()).isTrue();
+        verify(httpClient).post(anyString(), any(MultiMap.class), anyString(), anyLong());
+        verify(mockResponse).getStatusCode();
+        assertThat(capturedJsonNode).isEqualTo(expectedJsonNode);
+    }
+
+    @Test
+    public void shouldReceiveValidResponseOnAuctionContextForVideo() throws IOException {
+        // given
+        final Video video = setUpVideo();
+        final Imp imp = Imp.builder()
+                .id("adunitcodevalue")
+                .video(video)
+                .build();
+        final AuctionContext auctionContext = setupAuctionContext(imp);
+        event = AuctionEvent.builder()
+                .auctionContext(auctionContext)
+                .bidResponse(auctionContext.getBidResponse())
+                .build();
+
+        final HttpClientResponse mockResponse = mock(HttpClientResponse.class);
+        when(mockResponse.getStatusCode()).thenReturn(202);
+        when(httpClient.post(anyString(), any(MultiMap.class), jsonCaptor.capture(), anyLong()))
```

Comment:
the better practice is using a captor on the `verify` step

also, what about other parameters? headers, url, timeout?

Reply:
OK to follow the best practice, moving the captor to the `verify` step (and so, forget my previous comment).
However, here the test focuses on checking if the generated JSON payload match what is expected. I don't think that it's relevant to check other parameters here (or you would have to do it in all other tests, to stay consistent).
Maybe you can add a test to check if the configuration values are well parsed and forwarded to the `post` call (the url of the server, the timeout, and maybe the header).

Reply:
Have added jsonCaptor to verify `verify(httpClient).post(anyString(), any(MultiMap.class), jsonCaptor.capture(), anyLong());` 👍 

Still have to to keep capture in `when` step as it mocks `Future.succeededFuture(mockResponse)`. If I remove it then the `responseFuture` is null

Reply:
No. You can remove the `capture()` in the `when(...)` call. However, you have to re-order your assertions to do the `verify(..., jsonCaptor.capture(), ...)` before the JSON parsing.

```java
        when(httpClient.post(anyString(), any(MultiMap.class), anyString(), anyLong()))
                .thenReturn(Future.succeededFuture(mockResponse));

        final ObjectNode expectedJsonNode = (ObjectNode) mapper.readTree(parseJson(EXPECTED_JSON_PATH_BANNER));

        // when
        final Future<Void> result = target.processEvent(event);

        // then
        assertThat(result.succeeded()).isTrue();

        verify(httpClient).post(anyString(), any(MultiMap.class), jsonCaptor.capture(), anyLong());
        final ObjectNode capturedJsonNode = (ObjectNode) mapper.readTree(jsonCaptor.getValue());
        capturedJsonNode.put("greenbidsId", "testGreenbidsId");
        capturedJsonNode.put("billingId", "testBillingId");

        assertThat(capturedJsonNode).isEqualTo(expectedJsonNode);
```